### DISCUSSION TITLE
Jupyter Book and LaTeX

### DISCUSSION THREAD

**Discussion by chipbrock** [executablebooks/meta#511] [Q&A]
I've been writing HTML using Jupyter Books for months now and like it very much. I'm a refugee from RStudio and Bookdown. While as cumbersome as Bookdown is, I was able to fork outputs to either html5 and LaTeX with my own front-matter for the latter with ease. It is slick. But its support is way too limited to commit a working life to it. Jupyter seems much more robust now.

I am about to start a new project where the output needs to be LaTeX (I'd prefer that to pdf directly) with my own front matter. _What made the Bookdown learning doable was there were examples._ 

I cannot find examples on how to do this same thing in the Jupyter Book universe. Maybe I'm not looking in the right places, but the regular documentation is not helpful.

Does anyone know of an example setup to do Markdown -> Jupyter Book -> LaTeX? A simple working example would save me many hours of futzing around.

thanks....

**Comment by mmcky** [comment]
@chipbrock the `LaTeX`/`PDF` workflows for `jupyter-book` are powered by [sphinx-jupyterbook-latex](https://github.com/executablebooks/sphinx-jupyterbook-latex). The approach taken here (so far) is to work with the `sphinx.AST` and use transforms to get the `abstract syntax tree` in a form whereby we can continue to use the in-built sphinx writer (in general) and output nicer looking `pdf` files. 

One example were we use this extension to generate a `PDF` output is:

https://python.quantecon.org/intro.html

and you can download the pdf via the `Book (PDF)` button.

<img width="1677" alt="Screen Shot 2021-10-08 at 11 12 25 am" src="https://user-images.githubusercontent.com/8263752/136478624-c3104cb6-57e4-4c9c-9d83-476cc9b8685e.png">

This is largely taking an automated approach which focuses on `myst:md`/`rst`  -> `PDF` (via LaTeX).

One of the issues we face is the LaTeX output produced by Sphinx isn't particularly `human readable` as it wraps elements using `pygments` to syntax highlight code examples. I'd like to see some work go into making a more useful LaTeX intermediary file using some packages like `minted` for code (and letting LaTeX do the pygments steps).

I'd love to learn about your workflows to see how we can make improvements to `jupyter-book` workflows when working with LaTeX. 

Do you have any public examples of how you use metadata to control LaTeX output? 

To get LaTeX for jupyterbook you can use:

```bash
jb build <project> --builder=latex
```

with the results in `<project>/_build/latex`

and these [latex customizations via sphinx](https://www.sphinx-doc.org/en/master/latex.html) are also available via `_config.yml` and the [sphinx config](https://jupyterbook.org/advanced/sphinx.html?highlight=sphinx#manual-sphinx-configuration) passthrough.

**Reply by mmcky** [reply]
I have also opened an issue https://github.com/executablebooks/jupyter-book/issues/1489 to do pass at the docs with `LaTeX/PDF` in mind.

**Reply by chipbrock** [reply]
hi
Thanks for your reply - some comments in-line;

On Oct 7, 2021, at 8:17 PM, mmcky <notifications@github.com> wrote:


> @chipbrock the LaTeX/PDF workflows for jupyter-book are powered by sphinx-jupyterbook-latex. The approach taken here (so far) is to work with the sphinx.AST and use transforms to get the abstract syntax tree in a form whereby we can continue to use the in-built sphinx writer (in general) and output nicer looking pdf files.
> 
> One example were we use this extension to generate a PDF output is:
> 
> https://python.quantecon.org/intro.html
> 
> and you can download the pdf via the Book (PDF) button.



I see that. It looks nice, but what exactly did you do to make it happen? Step by step? How is front-matter with LaTeX package files and defined macros, etc included like a “regular” LaTeX experience? 

What this project overall lacks is some available tarball that prospective users could unravel and learn from. That’s precisely what the Bookdown guy did and it made everything possible with almost zero effort. 

https://bookdown.org/yihui/bookdown/get-started.html

If someone did that, it would be great and save hours and hours for potential users.

> This is largely taking an automated approach which focuses on myst:md/rst -> PDF (via LaTeX).
> 
> One of the issues we face is the LaTeX output produced by Sphinx isn't particularly human readable as it wraps elements using pygments to colours code for example. I'd like to see some work go into making a more useful LaTeX intermediary file using some packages like minted for code (and letting LaTeX do the pygments steps).


Yeah. That’s a little cumbersome. One would like to interact with the LaTeX source eventually. 

> I'd love to learn about your workflows to see how we can make improvements to jupyter-book workflows when working with LaTeX.



My workflow for Bookdown and Rstudio was to write in Rmarkdown, and when I was satisfied to 90%, then open inside of RStudio for the rest. One can control the LaTeX output in  the _output.yml file:

```
bookdown::gitbook:
  toc_depth: 1
  css: style.css
  config:
    toc:
      before: |
        <li><a href="./">toggle TOC with "s"</a></li>
      after: |
        <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
    download: ["pdf", "epub"]
bookdown::pdf_book:
  includes:
    in_header: preamble.tex
  latex_engine: xelatex
  citation_package: natbib
  keep_tex: yes
bookdown::epub_book: default
```
The preamble.tex file is just my personal front matter LaTeX stuff. The whole thing the just spits out. I learned to make figure commands be appropriate to html output or LaTeX output and the other would be ignored when not appropriate. So the markdown started to look ugly, but it could go either way.

So

```
<!--…pageimage……………………………….-->

```{r sunrise2, fig.cap="Sunrise or earthrise?", echo=FALSE,dev='png', eval=knitr::is_html_output()}
knitr::include_graphics('./images/0_Introduction/sunrise_H.jpg')
```


\begin{figure}
\includegraphics[width=1\linewidth]{./images/0_introduction/sunrise_L.jpg} \caption{Sunrise or earthrise?}\label{sunrise}
\end{figure}

```

Would go either way, but I'm not sure how I remember how I did that! I stopped using it as I really needed the html and was too fascinated with making it work.
